Create withReviews base HOC

Aljullu opened the pull request:
This PR creates a `withReviews` HOC so it can be used by `frontend-block.js` and `editor-block.js` as discussed in #841.

#### Accessibility

- [x] I've tested using only a keyboard (no mouse)
- [x] I've tested using a screen reader
- [x] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### How to test the changes in this Pull Request:

1. Create a post with a _Reviews by Product_ block and verify it loads reviews both in the editor and the frontend.
2. In the frontend, interact with the _Order by_ select and the _Load more_ button and verify they still work

nerrad commented on assets/js/base/hocs/with-reviews.js:
```diff
@@ -0,0 +1,150 @@
+			};
+		}
+
+		getReviews() {
```
It seems the only reason this is present is to pass it along as a way for wrapped components to trigger getting reviews.  Instead, I think you should update a `reviews` property in the state from any updates and only pass along the retrieved review data.  Passing along this function doesn't really decouple the data retrieval that much.  Ideally, the you want `withReviews` to handle all the retrieval and updating of reviews and just pass along the retrieved reviews to wrapped presentational components.  

For the `debounced` behaviour you have for the editor block implementation, you can control that via a `debounced` boolean prop fed into the hoc.  And that can control whether it does debounced getReviews on the `componentDidUpdate` lifecycle event (which should be in the hoc, not the presentational components).

Aljullu replied:
Right, the reason this is passed to the component was to simplify the frontend block so it didn't need the `debounce` dependency and didn't make any unncessary computations on `componentDidUpdate`.

`componentDidUpdate()` is just a couple of checks, so I don't think there is much performance difference, I moved it to the HOC as suggested. :+1: 

Regarding the debounce dependency, I moved it to `edit.js` and passed it as a prop. I would like to avoid loading dependencies which are not used in the frontend.

Let me know how it looks: 5d7bac1001ae56a3a0f73e32672a9c068a6115ae.

nerrad commented on assets/js/base/hocs/with-reviews.js:
```diff
@@ -0,0 +1,150 @@
+				onChangeArgs={ this.onChangeArgs }
+				isLoading={ loading }
+				reviews={ reviews }
+				totalReviews={ totalReviews }
```
The only thing that should be passed along here are the `reviews`, `totalReviews`, `error` and `isLoading` (along with any props you want to pass through).

I would also have the `LoadMoreButton` be something injected in the wrapped component via the `withReviews` hoc (overrideable via render prop).

Aljullu replied:
I think this is very similar to the discussion we had with the `withProduct()` HOC and the _Retry_ button, right?

If we move the `<LoadMoreButton>` from the block to the HOC, shouldn't we do the same with the `<ReviewOrderSelect>`, which triggers `onChangeOrderby` → `onChangeArgs` → `getReviews`? But that sounds like we would be moving presentational logic to the HOC. What if another component wanted to use `withReviews()` but instead of having a load more button it would want to use infinite scrolling? And what if instead of a select to change the order it wanted to have buttons? That would make our HOC unusable for those cases.

Aljullu commented on assets/js/base/hocs/with-reviews.js:
```diff
@@ -0,0 +1,150 @@
+			this.state = {
+				error: null,
+				loading: false,
+				overwrittenArgs: {},
```
I'm not sure to understand you with this. You mean moving the args to the state like this?
```ES6
this.state = {
	...
	queryArgs: {
		order,
		orderby,
		per_page: reviewsOnPageLoad,
	},
};
```
That means every time `attributes.orderby` is updated we would need to update the state to reflect the new order args, right? Same for the `reviewsOnPageLoad` and `productId` attributes. Isn't that a form of [derived state](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html)?

Aljullu commented:
@nerrad I refactored the HOC as we discussed yesterday in Slack. Feel free to take another look. :+1:

nerrad commented on assets/js/base/hocs/with-reviews.js:
```diff
@@ -0,0 +1,187 @@
+				...this.getArgs(),
+				per_page: reviewsToDisplay - reviews.length,
+				offset: reviews.length,
+			};
```
I wonder if we should just shift this to `updateListOfReviews`.  With the changes, it doesn't seem like `updateListOfReviews` needs to receive args as an argument anymore (nor reviews)?

nerrad commented on assets/js/base/hocs/with-reviews.js:
```diff
@@ -0,0 +1,187 @@
+			} );
+		}
+
+		updateListOfReviews( args, oldReviews = [] ) {
```
As mentioned in an earlier comment, it seems to me that there's no longer need for these arguments?

Aljullu replied:
I'm not sure to understand how the `oldReviews` param could be removed.

Aljullu commented on assets/js/base/hocs/with-reviews.js:
```diff
@@ -0,0 +1,187 @@
+				// it's better not to load the reviews immediately.
+				this.delayedAppendReviews();
+			} else if (
+				prevProps.categoryIds !== this.props.categoryIds ||
```
I refactored this a bit, `@wordpress/is-shallow-equal` would add another dependency so I tried to avoid it. Feel free to make any suggestions. :slightly_smiling_face:

Aljullu commented:
@nerrad thanks for the review! I addressed most of your feedback besides one suggestion where I added a question. :point_up_2: 

Besides that, this should be ready for another review.

nerrad commented on assets/js/base/hocs/with-reviews.js:
```diff
@@ -33,27 +31,50 @@ const withReviews = ( OriginalComponent ) => {
+				nextProps.categoryIds.join( ',' ) :
+				nextProps.categoryIds;
+
+			return prevCategoryIds !== nextCategoryIds;
```
All the category_id logic here would be much cleaner if you used `is-shallow-equal` :).  I saw your comment about not including it as a dependency but it's only [815B minified ](https://bundlephobia.com/result?p=@wordpress/is-shallow-equal@1.5.0), which is a worthwhile hit imo.

Aljullu replied:
I finally added it in c0fb984693dd09ccdfc840e0e8271ebbae19e225. My main concern was not about its size but that we don't bundle dependencies (so they can be shared between blocks/plugins), that means every new dependency we add produces another server request.

nerrad replied:
> we don't bundle dependencies (so they can be shared between blocks/plugins), that means every new dependency we add produces another server request.

Ya that is one downside.  However I'm not too worried about additional server request as long as:

- it's non-blocking (or _minimal_ blocking)
- it's small in size.

nerrad commented on assets/js/base/hocs/with-reviews.js:
```diff
@@ -88,30 +101,10 @@ const withReviews = ( OriginalComponent ) => {
 		}
 
-		updateListOfReviews( args, oldReviews = [] ) {
+		updateListOfReviews( oldReviews = [] ) {
```
Ya I see why you have this now, because replace reviews passes in no arg hence forcing a reset of the reviews listed 👍

nerrad commented:
hmm I approved this too soon, there's javascript tests failing (I'm guessing the new tests you added need updated).

Aljullu commented:
> hmm I approved this too soon, there's javascript tests failing (I'm guessing the new tests you added need updated).

Right, fixed it in 3ed94b29aed1e6c977cbf88c6a2d006063cc18ea.
